data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/leader/LeaderScheduler.java

```diff
@@ -0,0 +1,344 @@
+                        .collect(Collectors.toList());
+                LOG.info("Found {} root shards in total", rootShards.size());
+                rootShards.forEach(shard -> {
+                    createRootStreamPartition(tableInfo.getMetadata().getStreamArn(), shard, startTime, tableInfo.getMetadata().isExportRequired());
+                });
+                streamArns.add(tableInfo.getMetadata().getStreamArn());
+            }
+
+        });
+
+        LOG.debug("Update initialization state");
+        LeaderProgressState leaderProgressState = leaderPartition.getProgressState().get();
+        leaderProgressState.setStreamArns(streamArns);
+        leaderProgressState.setInitialized(true);
+        coordinator.saveProgressStateForPartition(leaderPartition, Duration.ofMinutes(DEFAULT_EXTEND_LEASE_MINUTES));
+    }
+
+
+    /**
+     * First run the shard discovery to detect new shards.
+     * Then find all the completed parent shards from the coordination store.
+     * The last step is to create a stream partition for each child shard been found.
+     */
+    private void discoveryAndCreateChildrenPartitions() {
+        LOG.info("Running shard discovery process to find child shards.");
+        long startTime = System.currentTimeMillis();
+        streamArns.forEach(streamArn -> {
+            shardManager.runDiscovery(streamArn);
+        });
+
+        List<EnhancedSourcePartition> sourcePartitions = coordinator.queryCompletedPartitions(StreamPartition.PARTITION_TYPE, Instant.now().minus(Duration.ofDays(1)));
+        LOG.info("Totally {} completed stream partitions in last day are found", sourcePartitions.size());
+
+        // Get the list of completed shard Ids.
+        List<String> completedShardIds = sourcePartitions.stream()
+                .map(sourcePartition -> ((StreamPartition) sourcePartition).getShardId())
+                .collect(Collectors.toList());
+        // Try to Create a stream partition for each child shards that have been found and not completed yet.
+        // If a shard is already created, it won't be created again.
+        sourcePartitions.forEach(sourcePartition -> {
+            StreamPartition streamPartition = (StreamPartition) sourcePartition;
+            List<String> childShardIds = shardManager.findChildShardIds(streamPartition.getStreamArn(), streamPartition.getShardId());
+            if (!childShardIds.isEmpty()) {
+                childShardIds.forEach(
+                        shardId -> {
+                            if (!completedShardIds.contains(shardId)) {
+                                createChildStreamPartition(streamPartition, shardId);
+                            }
+                        }
+                );
+            }
+        });
+        long endTime = System.currentTimeMillis();
+        LOG.info("Shard discovery process is completed, it took {} milliseconds in total", endTime - startTime);
```

Comment:
Should we report this as a metric instead of log it? Micrometer has a nice `timer()` that you can use.

https://github.com/opensearch-project/data-prepper/blob/d26748680468aa50bf1bf93c83cc005ae423a2e0/data-prepper-api/src/main/java/org/opensearch/dataprepper/metrics/PluginMetrics.java#L62-L64

Reply:
I put this as info log because I can't test the performance in my local, to call DDB api from my local, the latency is too high. So this process may take less than 1 second in OSIS v.s. many seconds in my local.

We will test the code, and if the performance is ok, we can change this to debug.  At least right now, this information is still for testing purpose, not sure if it has a production value, but can add it as a metric in the future.